File: server/src/domain.rs
use serde::{Deserialize, Serialize};
use time::Date;
use uuid::Uuid;

use crate::serde_iso_date;

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NewRecipe {
    pub name: String,
    pub quick: bool,
    pub body: String,
}

#[derive(Debug, Serialize, sqlx::FromRow)]
#[serde(rename_all = "camelCase")]
pub struct Recipe {
    pub id: Uuid,
    pub name: String,
    pub quick: bool,
    #[sqlx(rename = "body_html")]
    pub body: String,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Day {
    #[serde(with = "serde_iso_date")]
    pub date: Date,
    pub lunch: Meal,
    pub dinner: Meal,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
#[serde(tag = "type")]
pub enum Meal {
    Recipe(Recipe),
    Cheat,
    Unset,
}

pub enum MealType {
    Lunch,
    Dinner,
    Both,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NewFood {
    pub name: String,
    #[serde(with = "serde_iso_date")]
    pub best_before_date: Date,
}

#[derive(Serialize, sqlx::FromRow)]
#[serde(rename_all = "camelCase")]
pub struct Food {
    pub id: i32,
    pub name: String,
    #[serde(with = "serde_iso_date")]
    pub best_before_date: Date,
}

File: client/src/domain.ts
import { Dayjs } from "dayjs"

export interface LightRecipe {
  id: string
  name: string
  quick: boolean
}

export interface Recipe extends LightRecipe {
  body: string
}

export type NewRecipe = Omit<Recipe, "id">

export interface RecipeMeal extends LightRecipe {
  type: "recipe"
}

export interface CheatMeal {
  type: "cheat"
}

export interface UnsetMeal {
  type: "unset"
}

export type Meal = RecipeMeal | CheatMeal | UnsetMeal

export interface Day {
  date: Dayjs
  lunch: Meal
  dinner: Meal
}

export interface Food {
  id: number
  name: string
  bestBeforeDate: Dayjs
}

export interface NewFood {
  name: string
  bestBeforeDate: Dayjs
}
